Basic `vector` property tests

dmalkr opened the pull request:
Just copy `vector` package property tests

dmalkr committed:
Split vector space test into model and tests files

dmalkr committed:
Copy some property tests from `vector`

dmalkr committed:
Fix rebasing

Shimuuar commented on vecvec-lapack/vecvec-lapack.cabal:
```diff
@@ -44,20 +44,33 @@ Library
+                  RankNTypes
+                  ScopedTypeVariables
+                  TemplateHaskell
+                  TypeFamilies
```
Please enable CPP & TH on per module basis

dmalkr replied:
CPP & TH moved to modules.

dmalkr replied:
Both files, `Utilities.hs` and `TST/Property.hs` are copied from `vector`. And we hope tests from `vector` will be separated into other package. So we can just import them. So I didn't move any orphans, and just silent such warnings.

Shimuuar reviewed (changes requested):
Great! Thanks

There's quite a bit of orphans we don't need: all arbitrary instances for vectors from `vector`. Please kill unused ones and move remaining ones into `TST.Orphanage`. And drop `-fno-orphans` from cabal file

Please also make code warning free. `TST.Property` has tons of top-level functions without signature so just drop `{-# OPTIONS_GHC -Wno-missing-signatures #-}` there

Shimuuar commented:
Oh! Another thing. Please add possibility to generate vectors with non-1 stride in Arbitrary for Vec. See following as an example:

https://github.com/Shimuuar/vecvec/blob/7d73f0498328fad5a63f88762ef478feac179354/vecvec-lapack/tests/TST/Model.hs#L345-L348

dmalkr committed:
Remove TemplateHaskell, CPP extensions from cabal file

dmalkr committed:
WIP: tests for strided vectors

dmalkr committed:
Make strided vectors in `Arbitrary`

dmalkr commented:
> Oh! Another thing. Please add possibility to generate vectors with non-1 stride in Arbitrary for Vec. See following as an example:
> 
> https://github.com/Shimuuar/vecvec/blob/7d73f0498328fad5a63f88762ef478feac179354/vecvec-lapack/tests/TST/Model.hs#L345-L348

dmalkr commented:
Oh, got confused in the comment interface and accidentally closed the PR. :(

dmalkr commented:
> Oh! Another thing. Please add possibility to generate vectors with non-1 stride in Arbitrary for Vec. See following as an example:
> 
> https://github.com/Shimuuar/vecvec/blob/7d73f0498328fad5a63f88762ef478feac179354/vecvec-lapack/tests/TST/Model.hs#L345-L348

Done for `Arbitrary`. `TestData` also need to make `Vec`s from lists, but it have no access to random generating. So I made here fixed stride:

https://github.com/Shimuuar/vecvec/blob/e0fe928708bfa55e1f96a6a60ad8a2f7ab5b7963/vecvec-lapack/tests/Utilities.hs#L114-L120

dmalkr committed:
Remove warnings

Shimuuar committed:
Rework tests a bit.

We want to exercise both cases when stride=1 and >1.

 - Picking stride as positive is a bit dangerous. Underlying buffer size is
   stride × length so this choice may lead to excessive memory consumption.
   Furthermore we want to pick stride=1 with high probability

 - With TestData instance I'm not sure whether it makes sense to add stride un
   unmodel since (I think) it's only used in comparison. Here I pick stride
   using vector length. 777 was big overkill anyway

Effect is quite nice 26s test run time down to 15s
